### PR TITLE
Implement SDL-0187 Restructure iOS Threading and Queueing

### LINKED ISSUE #1028: [SDL 0187] Restructure iOS Threading and Queueing
Proposal: [Restructure iOS Threading and Queueing](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0187-restructure-ios-threading.md)

> This change targets the iOS library and restructures the existing threading / asynchronicity into a more manageable, consistent, and understandable system.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/543

Steering Committee Decision:

> The Steering Committee fully agreed to accept this proposal. It was also confirmed during discussion that this proposal does not make any changes regarding minimum version.

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #1028 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Unit tests updated and smoke tests added

### Summary
This PR restructures the threading in the library to use an "execution context" structure.

##### Enhancements
* Restructure the threading in the library to use an "execution context" structure.

### Tasks Remaining:
- [x] Restructure SDLFileManager tests

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLAudioStreamManager.m [hunk 1]
```diff
@@ -38,7 +39,7 @@ - (instancetype)initWithManager:(id<SDLStreamingAudioManagerType>)streamManager
     if (!self) { return nil; }
 
     _mutableQueue = [NSMutableArray array];
-    _audioQueue = dispatch_queue_create("com.sdl.audiomanager.transcode", DISPATCH_QUEUE_SERIAL);
+    _audioQueue = dispatch_queue_create_with_target("com.sdl.audiomanager.transcode", DISPATCH_QUEUE_SERIAL, [SDLGlobals sharedGlobals].sdlProcessingQueue);
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLAudioStreamManager.m]: When I install the library in a project I get this warning:
> 'dispatch_queue_create_with_target' is only available on iOS 10.0 or newer

When I click on the warning for more detail I get:

> 'dispatch_queue_create_with_target' has been marked as being introduced in iOS 10.0 here, but the deployment target is iOS 8.0.0

Comment (joeljfischer) [reply SmartDeviceLink/SDLAudioStreamManager.m]: It should be fixed now

**Comment by NicoleYarroch** [comment]
SYNC 3 doesn't seem to be handling multiple `Show` requests very well. Core either timeout after 20 seconds or response with a `GENERIC_ERROR` or `INVALID_DATA` request.

**Comment by joeljfischer** [comment]
After investigation, it seems that this is primarily a Ford issue with their HMI's handling of `Show` requests. Due to how the `ScreenManager`'s internal `TextAndGraphic` and `SoftButton` managers are structured, and how the example apps work, there are multiple show requests happening immediately when certain soft buttons are pressed, and pressing them over and over very quickly leads to errors from Ford's head unit.

I was unable to determine that this happens at a more frequent rate than the previous 6.3.1 release, and testing this branch with a full-featured internal media app showed qualitative improvement in speed when connected to Ford's head unit.

Additional improvements will likely only be possible by batching together soft button and graphic Show updates. This isn't currently possible with the structure of the `ScreenManager`'s internal managers, however, it may be possible once the `TextAndGraphic` sub-manager is restructured to use queues. This alone would be an improvement, and we could do further improvement by making the `SoftButtonManager` and `TextAndGraphic` queues combine updates when possible.

This would be quite a large bit of work, however, and is out of the scope of this proposal implementation, so I will leave that to future work.

**Pull request merged**